File: module-6-context/solutions/scratchpad.ts
/**
 * Module 6: Scratchpad tools
 * External memory via .scratchpad/{key}.md files.
 */
import * as fs from "fs/promises";
import * as path from "path";
import type { Tool } from "../../module-2-agent-loop/solutions/types.js";

export function createScratchpadTools(workspaceRoot: string): Tool[] {
  const scratchDir = path.join(workspaceRoot, ".scratchpad");

  const write: Tool = {
    name: "scratchpad-write",
    description: "Write a note to the scratchpad (persistent memory across turns)",
    input_schema: {
      type: "object" as const,
      properties: {
        key: { type: "string", description: "Note key (used as filename)" },
        content: { type: "string", description: "Note content (markdown)" },
      },
      required: ["key", "content"],
    },
    execute: async (input) => {
      await fs.mkdir(scratchDir, { recursive: true });
      const filePath = path.join(scratchDir, `${input.key}.md`);
      await fs.writeFile(filePath, String(input.content), "utf-8");
      return `Saved to scratchpad: ${input.key}`;
    },
  };

  const read: Tool = {
    name: "scratchpad-read",
    description: "Read a note from the scratchpad",
    input_schema: {
      type: "object" as const,
      properties: {
        key: { type: "string", description: "Note key to read" },
      },
      required: ["key"],
    },
    execute: async (input) => {
      const filePath = path.join(scratchDir, `${input.key}.md`);
      try {
        return await fs.readFile(filePath, "utf-8");
      } catch {
        return `No scratchpad entry found for key: ${input.key}`;
      }
    },
  };

  const list: Tool = {
    name: "scratchpad-list",
    description: "List all scratchpad notes",
    input_schema: { type: "object" as const, properties: {}, required: [] },
    execute: async () => {
      try {
        const entries = await fs.readdir(scratchDir);
        return entries.filter((e) => e.endsWith(".md")).map((e) => e.replace(".md", "")).join("\n") || "(empty)";
      } catch {
        return "(no scratchpad directory)";
      }
    },
  };

  return [write, read, list];
}

File: module-6-context/solutions/python/scratchpad.py
"""Module 6: Scratchpad tools."""
import os


def create_scratchpad_tools(workspace_root: str) -> list[dict]:
    scratch_dir = os.path.join(workspace_root, ".scratchpad")

    def write(inp):
        os.makedirs(scratch_dir, exist_ok=True)
        path = os.path.join(scratch_dir, f"{inp['key']}.md")
        with open(path, "w") as f:
            f.write(str(inp["content"]))
        return f"Saved to scratchpad: {inp['key']}"

    def read(inp):
        path = os.path.join(scratch_dir, f"{inp['key']}.md")
        try:
            with open(path) as f:
                return f.read()
        except FileNotFoundError:
            return f"No scratchpad entry found for key: {inp['key']}"

    def list_notes(_):
        try:
            entries = [e.replace(".md", "") for e in os.listdir(scratch_dir) if e.endswith(".md")]
            return "\n".join(entries) or "(empty)"
        except FileNotFoundError:
            return "(no scratchpad directory)"

    return [
        {"name": "scratchpad-write", "description": "Write a note to the scratchpad (persistent memory across turns)",
         "input_schema": {"type": "object", "properties": {"key": {"type": "string", "description": "Note key (used as filename)"}, "content": {"type": "string", "description": "Note content (markdown)"}}, "required": ["key", "content"]},
         "execute": write},
        {"name": "scratchpad-read", "description": "Read a note from the scratchpad",
         "input_schema": {"type": "object", "properties": {"key": {"type": "string", "description": "Note key to read"}}, "required": ["key"]},
         "execute": read},
        {"name": "scratchpad-list", "description": "List all scratchpad notes",
         "input_schema": {"type": "object", "properties": {}, "required": []},
         "execute": list_notes},
    ]
